### PR TITLE
Add com.surfshark.Surfshark

### PR DESCRIPTION
**Pull request by tomas0003**
### Please confirm your submission meets all the criteria

- [X] Please describe the application briefly.
   Surfshark VPN client with GUI.

- [X] The domain used for the application ID is [controlled by the application developers][appid-domain] and the [application id guidelines][appid] are followed.
- [X] I have read the and followed all the [Submission and licence requirements][reqs].
- [X] I have [built][build] and tested the submission locally.
- [X] I am an author/developer/upstream contributor of the project. If not, I contacted upstream developers about this submission.

[appid-domain]: https://docs.flathub.org/docs/for-app-authors/requirements/#control-over-domain-or-repository
[appid]: https://docs.flathub.org/docs/for-app-authors/requirements#application-id
[reqs]: https://docs.flathub.org/docs/for-app-authors/requirements
[build]: https://docs.flathub.org/docs/for-app-authors/submission/#before-submission

### CODE REVIEW

File: images/screenshot_main.png
Comment (bbhtt) [inline images/screenshot_main.png:1]: I don't see these screenshots being used?

File: libndp.yaml
Comment (bbhtt) [inline libndp.yaml:1]: Add this to the main manifest

File: libnma.yaml
Comment (bbhtt) [inline libnma.yaml:1]: Add this to the main manifest

File: network-manager-openvpn.yaml
Comment (bbhtt) [inline network-manager-openvpn.yaml:1]: Add this to the main manifest

File: network-manager.yaml
Comment (bbhtt) [inline network-manager.yaml:1]: Add this to the main manifest

File: com.surfshark.Surfshark.yaml [hunk 1]
```diff
@@ -0,0 +1,74 @@
+# yaml-language-server: $schema=https://raw.githubusercontent.com/flatpak/flatpak-builder/main/data/flatpak-manifest.schema.json
+
+app-id: com.surfshark.Surfshark
+runtime: org.gnome.Platform
```
Comment (bbhtt) [inline com.surfshark.Surfshark.yaml:4]: Does this actually need GNOME runtime or freedesktop will work too?

File: com.surfshark.Surfshark.yaml [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+runtime-version: '45'
```
Comment (bbhtt) [inline com.surfshark.Surfshark.yaml:5]:
```suggestion
runtime-version: '46'
```

46 is latest

File: com.surfshark.Surfshark.yaml [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+sdk: org.gnome.Sdk
+command: surfshark
+separate-locales: false
+finish-args:
+  - --socket=x11
+  - --share=ipc
+  - --share=network
+  - --system-talk-name=org.freedesktop.NetworkManager
+  - --talk-name=org.freedesktop.secrets
+  - --talk-name=org.freedesktop.Notifications
```
Comment (bbhtt) [inline com.surfshark.Surfshark.yaml:17]:
```suggestion
```

This goes through the portal.

File: com.surfshark.Surfshark.yaml [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+  - 'libnma.yaml'
+  - 'network-manager-openvpn.yaml'
+  - name: surfshark
+    buildsystem: simple
+    build-commands:
+      - ar -x surfshark_*.deb
+      - rm -f surfshark_*.deb
+      - tar -xf data.tar.xz
+
+      - install surfshark $FLATPAK_DEST/bin/
```
Comment (bbhtt) [inline com.surfshark.Surfshark.yaml:32]:
```suggestion
      - install -Dm0755 surfshark $FLATPAK_DEST/bin/surfshark
```

File: com.surfshark.Surfshark.yaml [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+      - install -D "usr/share/applications/surfshark.desktop" "${FLATPAK_DEST}/share/applications/${FLATPAK_ID}.desktop"
+      - install -D "usr/share/icons/hicolor/128x128/apps/surfshark.png" "${FLATPAK_DEST}/share/icons/hicolor/128x128/apps/${FLATPAK_ID}.png"
+      - install -D "usr/share/icons/hicolor/256x256/apps/surfshark.png" "${FLATPAK_DEST}/share/icons/hicolor/256x256/apps/${FLATPAK_ID}.png"
```
Comment (bbhtt) [inline com.surfshark.Surfshark.yaml:35]:
```suggestion
```

Only installing the largest one is fine.

**Comment by bbhtt** [comment]
> I am an author/developer/upstream contributor of the project. If not, I contacted upstream developers about this submission.

Did you contact Surfshark ?

File: com.surfshark.Surfshark.metainfo.xml [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+<component type="desktop-application">
+    <id>com.surfshark.Surfshark</id>
+
+    <name>Surfshark</name>
+    <summary>Connect to secure, private web!</summary>
+    <url type="homepage">https://surfshark.com/</url>
+    <url type="contact">https://support.surfshark.com/</url>
+
+    <metadata_license>CC-BY-SA-4.0</metadata_license>
+    <project_license>LicenseRef-proprietary=https://surfshark.com/terms-of-service</project_license>
```
Comment (bbhtt) [inline com.surfshark.Surfshark.metainfo.xml:11]: Does this allow redistribution? If not sources need to be extra-data.

File: images/screenshot_main.png
Comment (tomas0003) [reply images/screenshot_main.png:1]: added references via `https://raw.githubusercontent.com/flathub/com.surfshark.Surfshark`. URL should be valid once the build is published

File: com.surfshark.Surfshark.yaml (hunk 2 above)
Comment (tomas0003) [reply com.surfshark.Surfshark.yaml]: updated

File: com.surfshark.Surfshark.yaml (hunk 3 above)
Comment (tomas0003) [reply com.surfshark.Surfshark.yaml]: updated

File: com.surfshark.Surfshark.yaml (hunk 4 above)
Comment (tomas0003) [reply com.surfshark.Surfshark.yaml]: updated

File: com.surfshark.Surfshark.yaml (hunk 5 above)
Comment (tomas0003) [reply com.surfshark.Surfshark.yaml]: updated, 512x512 is the only exposed icon

File: com.surfshark.Surfshark.yaml [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+  - 'libndp.yaml'
+  - 'network-manager.yaml'
+  - 'libnma.yaml'
+  - 'network-manager-openvpn.yaml'
+  - name: surfshark
+    buildsystem: simple
+    build-commands:
+      - ar -x surfshark_*.deb
+      - rm -f surfshark_*.deb
+      - tar -xf data.tar.xz
```
[the remark this answers is not in the document]
Comment (tomas0003) [reply com.surfshark.Surfshark.yaml]: updated

File: com.surfshark.Surfshark.yaml [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+        commands:
+        - TMPDIR="$XDG_RUNTIME_DIR/app/$FLATPAK_ID" zypak-wrapper /app/opt/Surfshark/surfshark "$@"
+      - type: file
+        path: com.surfshark.Surfshark.metainfo.xml
+#
+      - type: file
+        only-arches:
+          - x86_64
+        url: https://ocean.surfshark.com/debian/pool/main/s/surfshark_3.0.3_amd64.deb
+        sha256: 534a46d0994f28cca0155efc625d949d810221d3d244681dc6b8f4b46b1efaf8
```
[the remark this answers is not in the document]
Comment (tomas0003) [reply com.surfshark.Surfshark.yaml]: refactored to utilisation of `extra-data`

File: com.surfshark.Surfshark.yaml [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+          type: anitya
+          project-id: 372017
+          is-main-source: true
+          url-template: https://ocean.surfshark.com/debian/pool/main/s/surfshark_${version}_amd64.deb
+      - type: file
+        only-arches:
+          - aarch64
+        url: https://ocean.surfshark.com/debian/pool/main/s/surfshark_3.0.3_arm64.deb
+        sha256: 7ed2100c19ddedc17ed059296a6ac602cf27b3d8bbf619d3235fd93c3e47a2bb
+        x-checker-data:
```
[the remark this answers is not in the document]
Comment (tomas0003) [reply com.surfshark.Surfshark.yaml]: refactored to utilisation of `extra-data`

File: com.surfshark.Surfshark.metainfo.xml (hunk 6 above)
Comment (tomas0003) [reply com.surfshark.Surfshark.metainfo.xml:11]: You are right, we would want to distribute the .deb application artifact ourselves. Refactored the manifest to utilise `extra-data`.

**Comment by tomas0003** [comment]
> Did you contact Surfshark ?
I am one of the project developers.

**Comment by bbhtt** [comment]
> > Did you contact Surfshark ?
> 
> I am one of the project developers, please find the organization mentioned in my profile

I think you need to make the membership public by going to the org settings page. I don't see anything on your profile.

Is this being submitted in official capacity? If so, I think it'd be fine to not use `extra-data`. Extra data is primarily when upstream does not allow redistribution, but if the developers are submitting it themselves, it is assumed that redistribution is allowed.

File: com.surfshark.Surfshark.metainfo.xml [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+            <caption>Surfshark interface</caption>
+            <image type="source">https://raw.githubusercontent.com/flathub/com.surfshark.Surfshark/master/screenshots/screenshot_main.png</image>
+        </screenshot>
+        <screenshot>
+            <caption>Static locations</caption>
+            <image type="source">https://raw.githubusercontent.com/flathub/com.surfshark.Surfshark/master/screenshots/screenshot_static_locations.png</image>
+        </screenshot>
+        <screenshot>
+            <caption>Multihop locations</caption>
+            <image type="source">https://raw.githubusercontent.com/flathub/com.surfshark.Surfshark/master/screenshots/screenshot_multihop_locations.png</image>
```
Comment (bbhtt) [inline com.surfshark.Surfshark.metainfo.xml:27]: We need working links for this for the build to pass. You can switch the URLs later when this is accepted.

For now you have to use https://raw.githubusercontent.com/flathub/flathub/42ae054b049c92575f1161302ecb220f235c4ec7/screenshots/screenshot_main.png etc.

File: com.surfshark.Surfshark.yaml [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+        commands:
+          - TMPDIR="$XDG_RUNTIME_DIR/app/$FLATPAK_ID" zypak-wrapper /app/extra/opt/Surfshark/surfshark "$@"
+      - type: script
+        dest-filename: apply_extra
+        commands:
+          - bsdtar -Oxf surfshark.deb 'data.tar.xz' | bsdtar -xf -
+          - cp -r usr/lib usr/share .
+          - rm -rf usr surfshark.deb
+#
+      - type: 'extra-data'
```
Comment (bbhtt) [inline com.surfshark.Surfshark.yaml:133]:
```suggestion

      - type: extra-data
```

File: com.surfshark.Surfshark.yaml [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+        only-arches:
+          - x86_64
+        url: https://ocean.surfshark.com/debian/pool/main/s/surfshark_3.0.4_amd64.deb
+        sha256: f7790097212def265b50ac8b51e3737710c846afb496cfb1a62610e708e9dc38
+        x-checker-data:
+          type: anitya
+          project-id: 372017
+          is-main-source: true
+          url-template: https://ocean.surfshark.com/debian/pool/main/s/surfshark_${version}_amd64.deb
+      - type: 'extra-data'
```
Comment (bbhtt) [inline com.surfshark.Surfshark.yaml:145]:
```suggestion
      - type: extra-data
```

File: surfshark.desktop [hunk 13]
```diff
@@ -0,0 +1,11 @@
+[Desktop Entry]
+Name=Surfshark
+Exec=/opt/Surfshark/surfshark %U
```
Comment (bbhtt) [inline surfshark.desktop:3]:
```suggestion
Exec=surfshark %U
```

**Comment by bbhtt** [comment]
Also if you are upstream I suggest shipping desktop file, icon and metainfo in the `.deb` directly. These aren't Flatpak or Flathub specific, the same files can be used by distros too.

File: com.surfshark.Surfshark.metainfo.xml (hunk 10 above)
Comment (tomas0003) [reply com.surfshark.Surfshark.metainfo.xml]: updated

**Comment by tomas0003** [comment]
> I think you need to make the membership public by going to the org settings page. I don't see anything on your profile.

Sorry, my bad. Should be seen now.

> Is this being submitted in official capacity? If so, I think it'd be fine to not use extra-data. Extra data is primarily when upstream does not allow redistribution, but if the developers are submitting it themselves, it is assumed that redistribution is allowed.

I see more reasons to use extra-data. This way we could control that deprecated and unsupported versions of app are not available via downgrading to certain flatpak repo commit. To be honest, I would leave the extra-data approach as is

**Comment by hfiguiere** [comment]
> I see more reasons to use extra-data. This way we could control that deprecated and unsupported versions of app are not available via downgrading to certain flatpak repo commit. To be honest, I would leave the extra-data approach as is

Don't do that. It is ANTI-user.

**Comment by bbhtt** [comment]
> I see more reasons to use extra-data. This way we could control that deprecated and unsupported versions of app are not available via downgrading to certain flatpak repo commit. 

The repo is pruned, old versions will not be left in the server for forever. In the unlikely case if surfshark's servers are blocked in someone's country the Flatpak will fail to install for them and it also makes debugging harder if say the upstream sources are rotated.

**Comment by tomas0003** [comment]
> The repo is pruned, old versions will not be left in the server for forever.

Okay, the oldest available commit I found for a random app in Flathub is from December 2023. Can I assume that old commits are kept for around 9 months? Thank you

File: com.surfshark.Surfshark.yaml (hunk 1 above)
Comment (tomas0003) [reply com.surfshark.Surfshark.yaml:4]: Added GNOME mainly for GObject introspection and GJS

**Comment by hfiguiere** [comment]
No you cannot assume this.

**Comment by bbhtt** [comment]
> Also if you are upstream I suggest shipping desktop file, icon and metainfo in the `.deb` directly. These aren't Flatpak or Flathub specific, the same files can be used by distros too.

Any answers to this? At least the screenshots should be in the surfshark website and not in the Flathub repo.

bot, build com.surfshark.Surfshark

File: com.surfshark.Surfshark.metainfo.xml [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+    <url type="contact">https://support.surfshark.com/</url>
+
+    <metadata_license>CC-BY-SA-4.0</metadata_license>
+    <project_license>LicenseRef-proprietary=https://surfshark.com/terms-of-service</project_license>
+    <developer id="com.surfshark">
+        <name>Surfshark B.V</name>
+    </developer>
+    <content_rating type="oars-1.1" />
+    <screenshots>
+        <screenshot>
```
Comment (bbhtt) [inline com.surfshark.Surfshark.metainfo.xml:17]:
```suggestion
        <screenshot type="default">
```

**Comment by bbhtt** [comment]
bot, build com.surfshark.Surfshark

**Comment by tomas0003** [comment]
looks good

**Comment by bbhtt** [comment]
bot, build com.surfshark.Surfshark

**Comment by bbhtt** [comment]
Seems to crash immediately in a VM:

```
#
# Fatal process out of memory: Failed to reserve virtual memory for CodeRange
#
----- Native stack trace -----

/app/bin/surfshark: line 2:     3 Trace/breakpoint trap   (core dumped) TMPDIR="$XDG_RUNTIME_DIR/app/$FLATPAK_ID" zypak-wrapper /app/opt/Surfshark/surfshark "$@"
```

**Comment by bbhtt** [comment]
Seems to work on host but weird that it says OOM. the VM has 6 GiB memory assigned to it and most of it is unused:

```
               total        used        free      shared  buff/cache   available
Mem:           5.8Gi       1.5Gi       2.8Gi       500Mi       2.2Gi       4.3Gi
Swap:          5.8Gi          0B       5.8Gi
```

**Pull request closed without merging**